grassroot-integration/src/main/java/za/org/grassroot/integration/location/LocationInfoBrokerImpl.java

```diff
@@ -367,6 +387,172 @@ public Province getProvinceFromGeoLocation(GeoLocation location) {
+
+        ParameterizedTypeReference<Map<String, Municipality>> responseType =
+                new ParameterizedTypeReference<Map<String, Municipality>>() {};
+
+        ResponseEntity<Map<String, Municipality>> result =
+                restTemplate.exchange(componentsBuilder.build().toUri(), HttpMethod.GET, null, responseType);
+
+        log.info("Received response: {}", result);
+
+        List<Municipality> municipalities = new ArrayList<>();
+        if(result.getBody() != null){
+           municipalities = new ArrayList<>(result.getBody().values());
+        }
+
+        log.info("Processed munis: {}", municipalities);
+
+        return municipalities;
+    }
+
+    @Override
+    public Municipality cacheMunicipalityByCoordinates(String userUid, double longitude, double latitude){
+        UriComponentsBuilder componentsBuilder = UriComponentsBuilder.fromHttpUrl("http://mapit.code4sa.org/point/4326/");
+        String latLong = longitude + "," + latitude;
+
+        componentsBuilder = componentsBuilder.path(latLong)
+                .queryParam("type","MN");
+
+        log.info("Composed URI: {}", componentsBuilder.toUriString());
+
+        ParameterizedTypeReference<Map<String,Municipality>> responseType =
+                new ParameterizedTypeReference<Map<String, Municipality>>() {};
+
+        ResponseEntity<Map<String,Municipality>> result =
+                restTemplate.exchange(componentsBuilder.build().toUri(),HttpMethod.GET,null,responseType);
+
+        Municipality municipality = null;
+        Cache cache = cacheManager.getCache("user_municipality");
+
+        if(result.getBody() != null){
+            municipality = result.getBody().entrySet().iterator().next().getValue();
+
+            cache.put(new Element(userUid,municipality));
+        }
+
+        log.info("Municipality = {}",municipality);
+
+        return municipality;
+    }
+//    Loading users with location not null
+    @Override
+    public  void cacheMunicipalitiesForUsersWithLocation(){
+        List<UserLocationLog> userLocationLogs = userLocationLogRepository.findAll();
+
+        Map<String,Municipality> userMunicipalityMap = new HashMap<>();
+
+        for(UserLocationLog userLocationLog:userLocationLogs){
+            Municipality municipality =
+                            cacheMunicipalityByCoordinates(userLocationLog.getUserUid(),userLocationLog.getLocation().getLongitude(),userLocationLog.getLocation().getLatitude());
+            userMunicipalityMap.put(userLocationLog.getUserUid(),municipality);
+        }
+    }
+
+    @Override
+    public Map<String,Municipality> getMunicipalitiesForUsersWithLocationFromCache(Set<String> userUids){
+        Cache cache = cacheManager.getCache("user_municipality");
+        Map<String,Municipality> municipalityMap = new HashMap<>();
+
+        List<String> cacheKeys = cache.getKeys();
+
+        for(String userUid: userUids){
+            if(cacheKeys.contains(userUid)){
```

Comment:
What if cache key does not contain user uid? Are we just leaving it out or do we fetch them?